c/tskit/trees.c

```diff
@@ -947,87 +984,1277 @@ tsk_treeseq_mean_descendants(tsk_treeseq_t *self,
+            goto out;
+        }
+    }
+out:
+    return ret;
+}
+
+typedef struct {
+    tsk_id_t *sample_sets;
+    tsk_size_t num_sample_sets;
+    tsk_size_t *sample_set_sizes;
+    tsk_id_t *set_indexes;
+} sample_count_stat_params_t;
+
+static int
+tsk_treeseq_sample_count_stat(tsk_treeseq_t *self,
+        tsk_size_t num_sample_sets, tsk_size_t *sample_set_sizes, tsk_id_t *sample_sets,
+        tsk_size_t result_dim, tsk_id_t *set_indexes, general_stat_func_t *f,
+        tsk_size_t num_windows, double *windows, double *result,
+        tsk_flags_t options)
+{
+    int ret = 0;
+    const size_t num_samples = self->num_samples;
+    const tsk_id_t num_nodes = (tsk_id_t) self->tables->nodes.num_rows;
+    size_t j, k, l;
+    tsk_id_t u, sample_index;
+    double *weights = NULL;
+    double *weight_row;
+    sample_count_stat_params_t args = {
+        .sample_sets = sample_sets,
+        .num_sample_sets = num_sample_sets,
+        .sample_set_sizes = sample_set_sizes,
+        .set_indexes = set_indexes
+    };
+
+    if (num_sample_sets == 0) {
+        ret = TSK_ERR_INSUFFICIENT_SAMPLE_SETS;
+        goto out;
+    }
+
+    weights = calloc(num_samples * num_sample_sets, sizeof(*weights));
+    if (weights == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+
+    j = 0;
+    for (k = 0; k < num_sample_sets; k++) {
+        if (sample_set_sizes[k] == 0) {
+            ret = TSK_ERR_EMPTY_SAMPLE_SET;
+            goto out;
+        }
+        for (l = 0; l < sample_set_sizes[k]; l++) {
+            u = sample_sets[j];
+            if (u < 0 || u >= num_nodes) {
+                ret = TSK_ERR_NODE_OUT_OF_BOUNDS;
+                goto out;
+            }
+            sample_index = self->sample_index_map[u];
+            if (sample_index == TSK_NULL) {
+                ret = TSK_ERR_BAD_SAMPLES;
+                goto out;
+            }
+            weight_row = GET_2D_ROW(weights, num_sample_sets, sample_index);
+            if (weight_row[k] != 0) {
+                ret = TSK_ERR_DUPLICATE_SAMPLE;
+                goto out;
+            }
+            weight_row[k] = 1;
+            j++;
+        }
+    }
+    ret =  tsk_treeseq_general_stat(self,
+        num_sample_sets, weights, result_dim, f, &args,
+        num_windows, windows, result, options);
+out:
+    tsk_safe_free(weights);
+    return ret;
+}
+
+/***********************************
+ * One way stats
+ ***********************************/
+
+static int
+diversity_summary_func(size_t state_dim, double *state, size_t TSK_UNUSED(result_dim),
+        double *result, void *params)
+{
+    sample_count_stat_params_t args = *(sample_count_stat_params_t *) params;
+    const double *x = state;
+    double n;
+    size_t j;
+
+    for (j = 0; j < state_dim; j++) {
+        n = (double) args.sample_set_sizes[j];
+        /* TODO: what happens when n is zero ? */
+        result[j] = x[j] * (n - x[j]) / (n * (n - 1));
```

Comment:
The denominator totally doesn't have to be in this function; and it seems like this is creating lots of extra multiplications and divisions. But I ran some (relatively small) checks and moving the `/ (n * (n - 1))` into `tsk_treeseq_diversity( )` had almost zero effect on run time. I'm surprised! I was all set to write a `column_normalize` function that we could use to do the normalizations, but perhaps we don't need to.

As for the `NaN`s, it looks like they are being correctly assigned. This is maybe [implementation dependent](https://www.gnu.org/software/libc/manual/html_node/Infinity-and-NaN.html). This is supposed to [throw an exception](https://www.gnu.org/software/libc/manual/html_node/FP-Exceptions.html#FP-Exceptions) that I guess we are not trapping.

Reply:
Each call to ``f``  currently does something like four mallocs and a two or three function calls. I'd imagine a function call costs at least 10 - 100 times as much as a division. A malloc surely costs 10 times as much again.

So, the divisions will almost certainly never be significant perf-wise. Perhaps there are numerical reasons for pulling them out? (Or conversely, for *not* pulling them out? Do we want to accumulate large values and divide afterward?)

Reply:
Since it's all an overall normalization, the order of operations shouldn't matter for floating point error, I think: (x/N - y/N) and (x - y)/N are the same.  Ok, we'll leave well enough alone.